bgzf/writer.go

```diff
@@ -13,24 +14,36 @@ import (
 // magicHexEOF is a particular empty bgzf block that marks the true EOF.
 var magicEOF []byte = makeMagicEOF()
 
-// Writer moves data -> compressor -> zipBlock -> file writer
-type Writer struct {
+// BlockWriter moves data -> compressor -> zipBlock -> file writer
+type BlockWriter struct {
 	w          io.Writer
 	compressor *gzip.Writer
 	zipBlock   *Block
 }
 
-// NewWriter creates a bgzf writer from any input writer.
-func NewWriter(w io.Writer) Writer {
-	var zw Writer
+type Writer struct {
+	bw  *BlockWriter
+	buf bytes.Buffer
+}
+
+// NewBlockWriter creates a bgzf writer from any input writer.
+func NewBlockWriter(w io.Writer) *BlockWriter {
+	var zw BlockWriter
 	zw.w = w
 	zw.zipBlock = NewBlock()
 	zw.compressor = gzip.NewWriter(zw.zipBlock)
-	return zw
+	return &zw
+}
+
+// NewWriter creates a bgzf writer that writes bytes instead of blocks
+func NewWriter(w io.Writer) *Writer {
+	var bw Writer
+	bw.bw = NewBlockWriter(w)
+	return &bw
 }
 
 // Write input bytes as a single bgzf block.
-func (w Writer) Write(p []byte) (n int, err error) {
+func (w *BlockWriter) Write(p []byte) (n int, err error) {
```

Comment:
Procedural possible?  I think procedural fits the gonomics style better, but not sure if this is to implement an interface somewhere